File: javascript/leetcode/easy/array/two-sum.js
/*
https://leetcode.com/problems/two-sum/
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
*/ 

//Brute Force
const twoSum = function(nums, target){
    const numsLength = nums.length;
    for(let i = 0; i < numsLength - 1; i++){
        const firstNum = nums[i];
        for(let j = i + 1; j < numsLength; j++){
            const secondNum = nums[j];
            if(firstNum + secondNum === target){
                return [i, j];
            }
        }
    }
};

//Hash Map
const twoSum = function(nums, target){
    const numsMap = new Map();
    
    for(let i = 0; i < nums.length; i++){
        const currVal = nums[i];
        const targetDiff = target - currVal;
        const mapValue = numsMap.get(targetDiff);

        if(mapValue !== undefined){
            return [mapValue, i];
        } else {
            numsMap.set(currVal, i);
        }
    }
};

File: java_lang/easy/TwoSum.java
package java_lang.easy;
import java.util.*; 
/*
https://leetcode.com/problems/two-sum/submissions/
*/ 

public class TwoSum {
    public static void main(String[] args){
        int[] testCaseArr = new int[]{2,7,11,15};
        int target = 9;
        System.out.println(Arrays.toString(twoSum(testCaseArr, target)));
    }

    public static int[] twoSum(int[] nums, int target){
        HashMap<Integer, Integer> numsMap = new HashMap<Integer, Integer>();
        int[] twoSumIndices = new int[2];

        for(int i = 0; i < nums.length; i++){
            int currNum = nums[i];
            int targetDiff = target - currNum;

            if(numsMap.containsKey(targetDiff)){
                twoSumIndices[0] = numsMap.get(targetDiff);
                twoSumIndices[1] = i;
                break;
            } else {
                numsMap.put(currNum, i);
            }
        }
        return twoSumIndices;
    }
}
